### DISCUSSION TITLE
Proposal: Better handling of secure transport

### DISCUSSION THREAD

**Discussion by helto4real** [vlang/v#8670] [Ideas]
# Better handling of secure transport
We need a better story for secure transport in the net module.  My proposal consist with two separate proposals that will make using a secure or non secure transport over a socket agnostic to the application using interfaces.

## Proposal 1: net.ssl module
Currently the implementation of secure transport using TLS etc is all over the place in vlib. We should merge all secure transport layer over sockets to one module like golang (tls lib). 
Possible solutions (first one is my proposal)
- net.ssl
- net.tls
- crypto.tls (like go)
- crypto.net 

If we make this in crypto module we can have everything crypto related in that module. I prefer `net.ssl` or `net.tls`though.

In NIX based os we just use openssl and on windows we use the schannel since openssl is a pain on windows to support. 

With the new net module built by @emily33901 we got a nice API with the TcpConn and UdpConn. I propose in the ssl module we introduce ether a `net.ssl.TcpConn` or `net.ssl.TlsConn` that provides the same interface as the net.TcpConn. The proposed interfaces in next proposal.

## Proposal 2: standard interfaces for reading, writing data on streams
With interfaces being quite stable now in V, we should, like golang, introduce reader, writer, listener interfaces etc. The benefit is obvious. We can use any kind of stream (socket, websocket, secure socket, secure websocket, string, file etc) implementing the same interface. This will also be a huge benefit for testing. Tests can use strings as they also implement the interfaces. As golang we put these interfaces in a new `io` module. `net` module is not good cause these interfaces can apply on other datatypes as strings for example.

The following interfaces suggested as start:

| Interface        | Expose function/s |
| ------------- | -------------- |
| io.Reader    | read(mut buf []byte) ?int<br />read_ptr(buf_ptr byteptr, len int) ?int|
| io.Writer      | write(bytes []byte) ?<br />write_ptr(b byteptr, len int) ?       |
| io.StringWriter | write_str(s string) ? |
| io.Closer      | close() ? |
| io.ReadWriter  | sum of Reader and Writer |
| io.ReadCloser  | sum of Reader and Closer |
| io.WriteCloser  | sum of Writer and Closer |
| io.ReadWriteCloser  | sum of Reader, Writer and Closer |

Other interfaces like ByteWriter, ByteReader, PipeReader, PipeWriter, Seeker, ReadAt, ReadFrom, RuneReader, RuneWriter could be implemented as needed later.

## Suggested path forward
The changes could be done in like suggested blocks to keep the PRs as small as possible

1. Move `net.openssl` to `net.ssl` _Edited after suggestion from @JalonSolov_ 
2. Make the net.ssl.TlsConn supporting openssl on all plattforms as first step
3. Implement the interfaces
4. Move x.websocket to net.websockets and use interfaces and the new net.ssl.TlsConn 
5. Implement `schannel`instead of openssl in the TlsConn for windows
6. ..n Rest of vlib that depends on sockets/secure sockets

Opinions? 
putting some names here that seem to have contributed in net space and are still active, please feel free to ping if I missed someone (sorry about that in that case).
@medvednikov @spytheman @joe-conigliaro @emily33901

**Comment by JalonSolov** [comment]
The only thing I have a problem with in the first proposal is the separation of `net.tls` and `net.ssl`.  First, `tls` is essentially `SSLv3`, but by having them in separate modules it implies they're different protocols/ways of doing things.

Also, V already has a precedent of "combining" some implementation details, such as handling both `http://` and `https://` in the single `net.http` module instead of having 2 separate modules.

For proposal 2:  I'm all for it!

I actually see these as 2 independent proposals, as one doesn't rely on the other.  It might be useful to split them into separate issues.

**Reply by helto4real** [reply]
@JalonSolov it is ether `net.tls` or `net.ssl`depending if we want to go the golang route or put it in net module. I vote for net module. The reason I do not want to put the ssl stuff directly in net module is not have to always have dependencies on openssl if not needed. And yes it will be split into several issues, wanted to show the bigger picture using interfaces like golang will definitely make it easier to be agnostic to secure layer or not.

**Comment by JalonSolov** [comment]
I'm perfectly fine with `net.ssl`.  I was just talking about having _both_ `net.ssl` and `net.tls` - I think that's _too much_ separation.

Not sure about `net.ssl.openssl` - again, the implementation specifics should be as unobtrusive as possible, to reduce confusion.  Which means there should just be `net.ssl`, and the code in there can use SChannel on Windows, OpenSSL on others, maybe LibreSSL on still others, or whatever else is needed.  To the regular V programmer, they just have to know to use `net.ssl` for secure stuff.

**Reply by helto4real** [reply]
Agree that is too much separation. That is why I said _ether_ one. And you are right we do not need submoduels. Thanks.

**Reply by JalonSolov** [reply]
> Agree that is too much separation. That is why I said _ether_ one. And you are right we do not need submoduels. Thanks.

My bad - I missed that it was either/or in the original post.  Went back and re-read, and... yep, my bad.  :-)

**Comment by dumblob** [comment]
Seeing the proposal (2), I like it on the first sight.

Maybe we could extend any `reader` interface(s) to support also `peek(...)` and `peek_ptr(...)` methods with the same signature as `read()` and `read_ptr()` to allow for increasing buffer size under the hood to allow zero-copy access in certain scenarios.

This is the major "flaw" I feel with usual "stream" interfaces - that they assume only `read & consume` use case, but there are many use cases where it's rather `read several times the same sequence or its parts and then stop reading altogether and forever` - i.e. no consuming (read *no copying*) needed at all.

**Reply by helto4real** [reply]
Yea good idea, that would be useful.
